lib/streamlit/elements/json.py

```diff
@@ -79,18 +79,18 @@ def json(
         """
         import streamlit as st
 
-        if isinstance(
-            body,
-            (
-                SessionStateProxy,
-                UserInfoProxy,
-                QueryParamsProxy,
-                StreamlitHeaders,
-                StreamlitCookies,
-            ),
-        ):
+        if is_custom_dict(body):
             body = body.to_dict()
 
+        if is_namedtuple(body):
+            body = body._asdict()
+
+        if isinstance(body, (map, enumerate)):
+            body = list(body)
+
+        if isinstance(body, (ChainMap, types.MappingProxyType, UserDict)):
+            body = dict(body)
```

Comment:
Could you please elaborate on what the `st.json` element has to do with the dataframe-handling support in this PR?

Reply:
This PR unifies all the data handling-related tests. This also includes tests with `st.write`, which uses `st.json` for some datastructures and `st.dataframe` for others. These data structures did not work correctly in `st.json` which would make the tests fail. Extracting this into a dedicated PR would require keeping some of the old data handling tests in.